File: Hashing /Question-1/Ques1Solution.cpp
#include<bits/stdc++.h>
using namespace std;

#define ll long long
#define nl '\n'

void solve(){
    long long n;
    cin >> n;
    ll m = 1 << n;
    ll arr[m];
    for(ll i = 0 ; i < m ; i++)
        cin >> arr[i];

    sort(arr , arr + m);

    map<ll,ll> sums;

    sums[0] = 1;

    vector<int> ans;

    vector<ll> subsums;

    for(int i = 0 ; i < m ; i++){
        if(sums[arr[i]]){
            sums[arr[i]]--;
        }
        else{
            ans.push_back(arr[i]);
            vector<ll> temp;
            for(auto&z : subsums)
                temp.push_back(z + arr[i]) , sums[z + arr[i]]++;

            for(auto&z : temp)
                subsums.push_back(z);
            subsums.push_back(arr[i]);
        }
    }

    for(auto&x : ans){
        cout << x << " ";
    }
    cout << nl;
}




int32_t main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int __;
    cin >> __;
    while(__--){
      solve();  

     }

  return 0;
}

File: Hashing /Question-1/Ques1Solution.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int t = sc.nextInt();
        while (t-- > 0) {
            solve(sc);
        }

        sc.close();
    }

    private static void solve(Scanner sc) {
        long n = sc.nextLong();
        int m = 1 << n;
        long[] arr = new long[m];

        for (int i = 0; i < m; i++) {
            arr[i] = sc.nextLong();
        }

        Arrays.sort(arr);

        Map<Long, Long> sums = new HashMap<>();
        sums.put(0L, 1L);

        List<Long> ans = new ArrayList<>();
        List<Long> subsums = new ArrayList<>();

        for (int i = 0; i < m; i++) {
            if (sums.getOrDefault(arr[i], 0L) > 0) {
                sums.put(arr[i], sums.get(arr[i]) - 1);
            } else {
                ans.add(arr[i]);
                List<Long> temp = new ArrayList<>();
                for (Long z : subsums) {
                    temp.add(z + arr[i]);
                    sums.put(z + arr[i], sums.getOrDefault(z + arr[i], 0L) + 1);
                }
                for (Long z : temp) {
                    subsums.add(z);
                }
                subsums.add(arr[i]);
            }
        }

        for (Long x : ans) {
            System.out.print(x + " ");
        }
        System.out.println();
    }
}
